### PR TITLE
Make `dora-rs` publishable on `crates.io`

### PR DESCRIPTION
**Pull request by haixuanTao**
This Pull Request simplify the dependency tree by:
- Removing unnecessary dependency with: `cargo +nightly udeps --all --all-targets`
- Centralize dependencies to the workspace declaration

It then adds a github job to publish most of dora-rs packages to `crates.io`.
Some packages can still not be published due to git dependencies. 

Blocked on:

- [x] https://github.com/elast0ny/raw_sync-rs/pull/29
- [x] https://github.com/getditto/safer_ffi/pull/155

### CODE REVIEW

**Comment by haixuanTao** [comment]
Just for info, I think that we can leave this one for the `0.2.1` release.

**Comment by phil-opp** [comment]
Sounds good! v

**Comment by phil-opp** [comment]
I updated the PR description with git dependencies that are still blocking this.

**Comment by phil-opp** [comment]
Looks like the `raw-sync` crate is really the last git dependency that we currently have. Unfortunately it's not clear if the crate is maintained anymore since there was no reply to the PR that we need. So maybe it makes sense to publish a fork to crates.io instead of waiting any longer...

**Comment by haixuanTao** [comment]
> Looks like the `raw-sync` crate is really the last git dependency that we currently have. Unfortunately it's not clear if the crate is maintained anymore since there was no reply to the PR that we need. So maybe it makes sense to publish a fork to crates.io instead of waiting any longer...

Yep, I'll do it. Just for info, i'm still debugging some issues with `cargo publish` https://github.com/dora-rs/dora/actions/workflows/release.yml . `Cargo publish` isolate the crates from the workspace which makes some workspace dependency surface, and block the publication.

**Comment by haixuanTao** [comment]
So, I finally managed to push all crates to `crates.io` as release candidates and fix all the issues. Mainly:
- Make sure to have license, description, and documentation page on all crates.
- Make sure to have the right dependency and that they are published with the right set of feature flags ( as they sometimes might inherit feature flags from the workspace )
- Make sure to publish the crates in the right order (outer leaf to inner leaf)

As you may now crates.io is quite picky.

You can find both:
- dora-node-api: dora-node-api
- dora-operator-api: dora-operator-api

for the Rust API.

As well as:
- dora-runtime: dora-runtime
- dora-daemon: dora-daemon
- dora-coordinator: dora-coordinator
- dora-cli: dora-cli

They should be installable via:
```bash
cargo install dora-cli
```bash

File: Cargo.lock [hunk 1]
```diff
@@ -250,7 +250,7 @@ checksum = "8c3c1a368f70d6cf7302d78f8f7093da241fb8e8807c05cc9e51a125895a6d5b"
 
 [[package]]
 name = "benchmark-example-node"
-version = "0.2.0"
+version = "0.2.0-rc-10"
```
Comment (phil-opp) [inline Cargo.lock:253]: These should match the versions specified in the Cargo.toml. I think you accidentally commited your lockfile after doing the test release?

**Review by phil-opp — COMMENTED** [review]
Thanks a lot for all your work on this! Looks all good to me, there are just two small things.

File: binaries/cli/src/template/c/mod.rs [hunk 2]
```diff
@@ -53,8 +53,8 @@ fn create_dataflow(name: String, path: Option<PathBuf>) -> Result<(), eyre::ErrR
 
 fn create_operator(name: String, path: Option<PathBuf>) -> Result<(), eyre::ErrReport> {
     const OPERATOR: &str = include_str!("operator/operator-template.c");
-    const HEADER_API: &str = include_str!("../../../../../apis/c/operator/operator_api.h");
-    const HEADER_TYPE: &str = include_str!("../../../../../apis/c/operator/operator_types.h");
+    const HEADER_API: &str = include_str!("operator/operator_api.h");
+    const HEADER_TYPE: &str = include_str!("operator/operator_types.h");
```
[the remark this answers is not in the document]
Comment (haixuanTao) [reply binaries/cli/src/template/c/mod.rs:57]: I see make sense! Sure can do that!

File: Cargo.lock (hunk 1 above)
Comment (haixuanTao) [reply Cargo.lock:253]: Indeed. Good catch!

File: apis/c/operator/src/lib.rs [hunk 3]
```diff
@@ -1 +1,2 @@
-
+pub static HEADER_OPERATOR_API: &str = include_str!("../operator_api.h");
+pub static HEADER_OPERATOR_TYPES: &str = include_str!("../operator_types.h");
```
[the remark this answers is not in the document]
Comment (haixuanTao) [reply apis/c/operator/src/lib.rs:2]: Indeed.

**Review by phil-opp — APPROVED** [review]
Thanks!

**Pull request merged**